File: src/10.graph/36.417.go
package graph

func PacificAtlantic(heights [][]int) [][]int {
	m := len(heights)
	n := len(heights[0])
	ans := [][]int{}
	po := make([][]int, m)
	ao := make([][]int, m)
	for i := range po {
		po[i] = make([]int, n)
		ao[i] = make([]int, n)
	}

	var dfs func(r, c int, oo [][]int)

	dfs = func(r, c int, oo [][]int) {
		if oo[r][c] == 1 {
			return
		}
		oo[r][c] = 1

		if r-1 >= 0 && heights[r-1][c] >= heights[r][c] {
			dfs(r-1, c, oo)
		}

		if r+1 < m && heights[r+1][c] >= heights[r][c] {
			dfs(r+1, c, oo)
		}

		if c-1 >= 0 && heights[r][c-1] >= heights[r][c] {
			dfs(r, c-1, oo)
		}

		if c+1 < n && heights[r][c+1] >= heights[r][c] {
			dfs(r, c+1, oo)
		}
	}

	// 左右两边
	for i := 0; i < m; i++ {
		dfs(i, 0, po)
		dfs(i, n-1, ao)
	}

	// 上下两边
	for j := 0; j < n; j++ {
		dfs(0, j, po)
		dfs(m-1, j, ao)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if po[i][j] == 1 && ao[i][j] == 1 {
				ans = append(ans, []int{i, j})
			}
		}
	}

	return ans
}

File: src/10.graph/36.417.ts
export function pacificAtlantic(heights: number[][]): number[][] {
    const m = heights.length;
    const n = heights[0].length;
    const po = new Array<boolean[][]>(m)
        .fill([])
        .map(() => new Array<boolean>(n).fill(false));

    const ao = new Array<boolean[][]>(m)
        .fill([])
        .map(() => new Array(n).fill(false));

    const ans: number[][] = [];

    const dfs = (r: number, c: number, oo: boolean[][]) => {
        if (oo[r][c]) {
            return;
        }
        oo[r][c] = true;

        if (r - 1 >= 0 && heights[r - 1][c] >= heights[r][c]) {
            dfs(r - 1, c, oo);
        }

        if (r + 1 < m && heights[r + 1][c] >= heights[r][c]) {
            dfs(r + 1, c, oo);
        }

        if (c - 1 >= 0 && heights[r][c - 1] >= heights[r][c]) {
            dfs(r, c - 1, oo);
        }

        if (c + 1 < n && heights[r][c + 1] >= heights[r][c]) {
            dfs(r, c + 1, oo);
        }
    };

    for (let i = 0; i < m; i++) {
        // 左边界
        dfs(i, 0, po);
        // 右边界
        dfs(i, n - 1, ao);
    }

    for (let j = 0; j < n; j++) {
        // 上边界
        dfs(0, j, po);
        // 下边界
        dfs(m - 1, j, ao);
    }

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (po[i][j] && ao[i][j]) {
                ans.push([i, j]);
            }
        }
    }

    return ans;
}
